seaborn/matrix.py

```diff
@@ -522,45 +533,22 @@ def __init__(self, data, pivot_kws=None, z_score=None, standard_scale=None,
         self.fig = plt.figure(figsize=figsize)
 
         if row_colors is not None:
-            row_colors = _convert_colors(row_colors)
+            row_colors = _convert_colors_frame(row_colors)
         self.row_colors = row_colors
         if col_colors is not None:
-            col_colors = _convert_colors(col_colors)
+            col_colors = _convert_colors_frame(col_colors)
         self.col_colors = col_colors
 
-        width_ratios = self.dim_ratios(self.row_colors,
-                                       figsize=figsize,
-                                       axis=1)
-
-        height_ratios = self.dim_ratios(self.col_colors,
-                                        figsize=figsize,
-                                        axis=0)
-        nrows = 3 if self.col_colors is None else 4
-        ncols = 3 if self.row_colors is None else 4
-
-        self.gs = gridspec.GridSpec(nrows, ncols, wspace=0.01, hspace=0.01,
-                                    width_ratios=width_ratios,
-                                    height_ratios=height_ratios)
-
-        self.ax_row_dendrogram = self.fig.add_subplot(self.gs[nrows - 1, 0:2],
-                                                      axisbg="white")
-        self.ax_col_dendrogram = self.fig.add_subplot(self.gs[0:2, ncols - 1],
-                                                      axisbg="white")
+        # Setup axes for current configuration.
+        axes = self.setup_axes(self.fig, figsize,
+                               row_colors, col_colors,
+                               row_cluster, col_cluster,
+                               col_colors_ratio=col_colors_ratio,
+                               row_colors_ratio=row_colors_ratio)
 
-        self.ax_row_colors = None
-        self.ax_col_colors = None
-
-        if self.row_colors is not None:
-            self.ax_row_colors = self.fig.add_subplot(
-                self.gs[nrows - 1, ncols - 2])
-        if self.col_colors is not None:
-            self.ax_col_colors = self.fig.add_subplot(
-                self.gs[nrows - 2, ncols - 1])
-
-        self.ax_heatmap = self.fig.add_subplot(self.gs[nrows - 1, ncols - 1])
-
-        # colorbar for scale to left corner
-        self.cax = self.fig.add_subplot(self.gs[0, 0])
+        self.ax_row_dendrogram, self.ax_col_dendrogram, \
```

Comment:
I personally would think the `self.ax_row_dendrogram` (and similar) assignments should be done inside the `self.setup_axes()` function: `setup_axes` name kind of implies this, and if we ever wanted to change something, we would only need to change it in one place. The code will look cleaner too.